--- phoebe/src/phoebe/PhoebeCanvas.java.old	2006-01-13 12:26:34.000000000 +0000
+++ phoebe/src/phoebe/PhoebeCanvas.java	2006-01-10 15:32:13.000000000 +0000
@@ -74,10 +74,14 @@
 	public PhoebeCanvas ()
 	{
 		super();
+		if (GraphicsEnvironment.isHeadless()) {
+			// do not add DropTarget
+			dropTarget = null;
+		} else {
 		dropTarget = new DropTarget(this, // component
 				DnDConstants.ACTION_COPY, // actions
 				this); // DropTargetListener
-
+		}
 	}
 	
     /**
