*** src/net/sf/antcontrib/logic/OutOfDate.java.old      Fri May  6 15:17:12 2005
--- src/net/sf/antcontrib/logic/OutOfDate.java  Fri May  6 14:08:00 2005
***************
*** 320,325 ****
--- 320,328 ----
                      ret = true;
                  }
              }
+             if (paths.length == 0) {
+                ret = true;
+             }
          }
  
          // Mapper Paths
          